Add HomeKit Controller heater-cooler devices

vfreex opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
Some new HomeKit climate devices, like [XiaoMi  Aqara Air Conditioning Controller P3][1] are heater-cooler devices instead of
thermostat devices. They work fine when pairing with iPhone but are not shown in home-assistant homekit controller integration.

This PR adds heater-cooler devices to HomeKit Controller.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [x] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->
- Requires https://github.com/Jc2k/aiohomekit/pull/18 to function.
- Tested with my XiaoMi Air Conditioning Controller P3 but not other devices:
<img width="684" alt="Screen Shot 2020-08-18 at 2 31 48 AM" src="https://user-images.githubusercontent.com/9941773/90435654-da2fbd00-e101-11ea-9e6e-77faa9e24233.png">
<img width="442" alt="Screen Shot 2020-08-18 at 2 32 00 AM" src="https://user-images.githubusercontent.com/9941773/90435657-dbf98080-e101-11ea-8382-8e3b7077f038.png">
<img width="1653" alt="Screen Shot 2020-08-18 at 2 34 10 AM" src="https://user-images.githubusercontent.com/9941773/90435659-dc921700-e101-11ea-95ef-44e876cd081d.png">

- This PR fixes or closes issue: fixes https://github.com/home-assistant/core/issues/30384


## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-asc+-review%3Aapproved

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

[1]: https://www.aqara.com/cn/productDetail/d33

vfreex committed:
HomeKit Controller: Add support for heater-cooler devices

Some new HomeKit climate devices, like [XiaoMi Air Conditioning Controller P3][1] are heater-cooler devices rather than
thermostat devices. They work fine when pairing with iPhone but are not shown in home-assistant homekit controller integration.

This PR adds support for heater-cooler devices. Tested with my XiaoMi Air Conditioning Controller P3 but not other devices.

[1]: https://www.aqara.com/cn/productDetail/d33

homeassistant commented:
Hi @vfreex, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#38979).

Once you do that we will be able to review and accept this pull request.

Thanks!

Jc2k commented on homeassistant/components/homekit_controller/climate.py:
```diff
@@ -56,15 +89,219 @@ async def async_setup_entry(hass, config_entry, async_add_entities):
         info = {"aid": aid, "iid": service["iid"]}
-        async_add_entities([HomeKitClimateEntity(conn, info)], True)
-        return True
+        if service["stype"] == "thermostat":
```
Can you use a dict here like the other entity types? E.g. https://github.com/home-assistant/core/blob/dev/homeassistant/components/homekit_controller/binary_sensor.py#L120

Jc2k commented on homeassistant/components/homekit_controller/climate.py:
```diff
@@ -56,15 +89,219 @@ async def async_setup_entry(hass, config_entry, async_add_entities):
+        char = self.service[CharacteristicsTypes.SWING_MODE]
+        return [
+            SWING_MODE_HOMEKIT_TO_HASS[swing]
+            for swing in range(char.minValue, char.maxValue + 1)
```
You need to consider valid_values and not just minValue and maxValue. Manufactures can (and do) use both. clamp_enum_to_char may help here. There is an example in this file.

Jc2k commented on homeassistant/components/homekit_controller/climate.py:
```diff
@@ -56,15 +89,219 @@ async def async_setup_entry(hass, config_entry, async_add_entities):
+        # CCan be 0 - 2 (Auto, Heat, Cool)
+        if (
+            self.service.has(CharacteristicsTypes.ACTIVE)
+            and self.service.value(CharacteristicsTypes.ACTIVE)
```
`self.service.value` returns None if the characterstic is not present (or you can provide a default value) - so you can get rid of this `has` check here.

Jc2k commented on homeassistant/components/homekit_controller/climate.py:
```diff
@@ -56,15 +89,219 @@ async def async_setup_entry(hass, config_entry, async_add_entities):
+    def target_temperature(self):
+        """Return the temperature we try to reach."""
+        state = self.service.value(CharacteristicsTypes.TARGET_HEATER_COOLER_STATE)
+        if state == TargetHeaterCoolerStateValues.COOL and self.service.has(
```
You can do something like:

```
if state == TargetHeaterCoolerStateValues.COOL:
    return self.service.value(CharacteristicsTypes.TEMPERATURE_COOLING_THRESHOLD)
```

It will return None if the char isn't present, which is what happens in your current implementation anyway.

Jc2k commented on homeassistant/components/homekit_controller/climate.py:
```diff
@@ -56,15 +89,219 @@ async def async_setup_entry(hass, config_entry, async_add_entities):
+        char = self.service[CharacteristicsTypes.TARGET_HEATER_COOLER_STATE]
+        ret = [
+            TARGET_HEATER_COOLER_STATE_HOMEKIT_TO_HASS[state]
+            for state in range(char.minValue, char.maxValue + 1)
```
You should consider valid_values as well as min/max value for enum fields. See my swing_modes comment.

Jc2k reviewed (changes requested):
Nice!

I've released 0.2.49 with your aiohomekit changes in. Please update homekit_controller's `mainfest.json` and then run `./script/gen_requirements_all.py` to update the requirement files.

Tests are a requirement for homekit_controller. Our coverage is 100% for some modules (including climate). Especially as I don't have a device of this type it would be very easy to introduce a regression. The current climate tests [here](https://github.com/home-assistant/core/blob/dev/tests/components/homekit_controller/test_climate.py).

Jc2k commented:
Also i noticed you did a force push - when a review has started you can rebase your changes to pick up changes from `dev`, but don't edit or squash commits as it makes the review harder. When it's merged we'll squash it ourselves, so don't worry about number of commits or anything. Cheers!

vfreex committed:
apply review suggestions

vfreex commented:
@Jc2k Thanks for your suggestions! I've pushed a new commit to address the issues and added unit tests. Could you have a look again?

vfreex committed:
Run isort and flake8

Jc2k commented on homeassistant/components/homekit_controller/climate.py:
```diff
@@ -56,15 +89,218 @@ async def async_setup_entry(hass, config_entry, async_add_entities):
+        modes = [
+            TARGET_HEATER_COOLER_STATE_HOMEKIT_TO_HASS[mode] for mode in valid_values
+        ]
+        if self.service.has(CharacteristicsTypes.ACTIVE):
```
`ACTIVE` is a required characteristic of the heater coler service, so this check isn't required.

vfreex replied:
ok. next push.

Jc2k reviewed (changes requested):
Thank you for the tests :-)

What happens when a device is in `AUTO` mode? There are a couple of places that check `TARGET_HEATER_COOLER_STATE` for `HEAT` or `COOL` but don't handle `AUTO`.

Do you think if the device supports `AUTO` then the integration should support `SUPPORT_TARGET_TEMPERATURE_RANGE`? That enables you to see both temperatures at once in the HA UI, and to set both at once.

vfreex commented:
@Jc2k 
> What happens when a device is in `AUTO` mode? There are a couple of places that check `TARGET_HEATER_COOLER_STATE` for `HEAT` or `COOL` but don't handle `AUTO`.

I am not sure what will happen if a device is in `AUTO` mode. My device doesn't support AUTO mode with HomeKit protocol, so I can't test it nor find any official document about the behavior on the internet. Following is the behavior of my device during my test:

- When directly paring with iPhone with HomeKit, it only shows `Cool`, `Heat`, and `Off` options.
- It's summer in my country. If I change my air conditioner to auto mode with remote control or Aqara app, the temperature will disappear on air conditioner's display and Aqara app. Both `TARGET_HEATER_COOLER_STATE` and `TARGET_HEATER_COOLER_STATE` will report `cool/cooling`. But I am not sure what will happen in winter.
- In AUTO mode, `TEMPERATURE_HEATING_THRESHOLD` will report 22C and the `TEMPERATURE_COOLING_THRESHOLD` will return 26C.
- In AUTO mode, set `TEMPERATURE_HEATING_THRESHOLD` or `TEMPERATURE_COOLING_THRESHOLD` has no effects (device doesn't respond)

> Do you think if the device supports `AUTO` then the integration should support `SUPPORT_TARGET_TEMPERATURE_RANGE`? That enables you to see both temperatures at once in the HA UI, and to set both at once.

My device doesn't support `SUPPORT_TARGET_TEMPERATURE_RANGE` and setting `TEMPERATURE_HEATING_THRESHOLD` or `TEMPERATURE_COOLING_THRESHOLD` has no effects. So I don't want to make such an assumption until we see a real device supports this.

So I'd like to hold off auto mode temperature control until we see a real device supports this. However I don't know how to disallow users to set the temperature when AUTO mode is engaged.

What do you think?

vfreex committed:
Remove ACTIVE characteristic check

It is required.

Jc2k commented:
IMO the R2 spec is vaguely worded about this, but the way I read it is that in HomeKit AUTO mode if the temperature goes below the heating threshold temperature, it turns the heating on until its above that temperature. And if the temperature is above the cooling threshold temperature, it cools. Which means in AUTO mode you need to be able to set both characteristics, which is why I think `SUPPORT_TARGET_TEMPERATURE_RANGE` makes sense for those devices - it means set_temperature is potentially called with both values.

As your device can't be used to test this, I think that is is sensible to proceed as you suggest. Maybe `set_temperature` should emit a warning into the logs if the device is in `AUTO` mode in that case? And maybe the same for `set_hvac_mode`. Maybe it could say to raise a ticket?

One other thing before i :+1: this, I wonder if `TargetHeaterCoolerStateValues.AUTOMATIC` actually maps to `HVAC_MODE_HEAT_COOL`? In HA, `AUTO` means "The device is set to a schedule, learned behavior, AI.". But `HEAT_COOL` means "The device supports heating/cooling to a range", which I think is closer to what the spec says. What do you think?

vfreex committed:
Emit a warning when setting temperature in AUTO (and other) mode.

vfreex commented:
Yeah, mapping `TargetHeaterCoolerStateValues.AUTOMATIC` to `HVAC_MODE_HEAT_COOL` sounds more reasonable based on the assumption.

Jc2k reviewed (approved):
OK, I am happy with this from a HomeKit controller POV, so :+1: from me. Thanks for the contribution!

vfreex committed:
heater-cooler hvac_action should always return off when turned off

I (and another user) found an issue: When a heater-cooler device is turned off, Home Assistant still thinks the device is heating/cooling
and associated automation is broken. This might be vendor specific: Even though the device's `ACTIVE` char reports `INACTIVE`, the `CURRENT_HEATER_COOLER_STATE` char still reports `HEATING`/`COOLING`.

This PR is trying to match Home app's behavior: `hvac_action` should always return `off` when `ACTIVE` char is `INACTIVE` no mater whatever value `CURRENT_HEATER_COOLER_STATE` reports.
